packages/ruleset/src/functions/use-date-based-format.js

```diff
@@ -0,0 +1,401 @@
+            '.'
+          )}" is a string or integer`
+        );
+
+        // If the schema is determined to be a date-time schema by the name alone,
+        // we can return - no need to look for an example value.
+        return [
+          {
+            message:
+              'According to its name, this property should use type "string" and format "date" or "date-time"',
+            path,
+          },
+        ];
+      }
+    }
+  }
+
+  // Check example values for string schemas.
+  if (isStringSchema(schema)) {
+    // Search the list of examples for an example object that contains a value
+    // for this property.
+    const exampleValue = findExample(logicalPath, examples, indirectExamples);
+    if (exampleValue !== undefined) {
+      logger.debug(
+        `${ruleId}: example value found for string schema at location ${path.join(
+          '.'
+        )}: ${exampleValue}`
+      );
+
+      if (isDateBasedValue(exampleValue)) {
+        return [
+          {
+            message:
+              'According to its example value, this schema should use type "string" and format "date" or "date-time"',
+            path,
+          },
+        ];
+      }
+    }
+  }
+
+  return [];
+}
+
+// This function checks all of the examples we've gathered while processing
+// schemas to check if once of them defines a value for the specific property
+// or string schema that we are looking at. It returns the first value found.
+function findExample(logicalPath, examples, indirectExamples) {
+  // First, look at the indirect examples (those included through OpenAPI
+  // fields) as they are given priority by OpenAPI over native schema examples.
+  if (indirectExamples.length) {
+    logger.debug(`${ruleId}: Looking for value within indirect examples.`);
+    const value = findValueInExamples(
+      indirectExamples,
+      logicalPath,
+      // Indirect examples, like primary schemas, need an empty path array.
+      []
+    );
+
+    // If we find a value, go ahead and return.
+    if (value !== undefined) {
+      logger.debug(
+        `${ruleId}: example value found in indirect example: ${value}`
+      );
+
+      return value;
+    }
+  }
+
+  // Look for examples at different paths up the logical path.
+  const examplePath = [...logicalPath];
+
+  // Look at each level of hierarchy of the example path, INCLUDING examples on
+  // primary schemas, which will have an example path of an empty array. For
+  // that reason, we can't loop through the example path elements, as we would
+  // miss the iteration with no elements in the array. Using a simple for-loop
+  // that executes one more time than there are segments in the example path,
+  // we execute this logic for the empty path scenario as well, and thereby
+  // properly handle primary schemas. The length of the example path is used to
+  // initialize the counter, rather than in the conditional, since the array
+  // will change length as we progress.
```

Comment:
Per the comment above, you probably don't need this at all, but as a matter of edification, I think `do-while` is much clearer here:

```
do {
  ...
} while (examplePath.pop() !== undefined);
```

Reply:
Oh, 100%. I feel silly haha - I had a `do-while` loop initially but ran into the issue of executing the code for the empty array, and didn't even consider popping the array item within the conditional itself! I knew there was a better way to do this 😅